packages/hooks/src/useCollectionQuery/useCollectionQuery.ts

```diff
@@ -0,0 +1,343 @@
+  TQuery,
+  TSubscription
+>): CollectionQueryResult<TQuery> {
+  const {
+    data,
+    loading,
+    refetch,
+    error,
+    fetchMore,
+    subscribeToMore,
+  } = useQuery<TQuery>(query, queryOptions);
+
+  const isMounted = useIsMounted();
+  const [loadingRefresh, setLoadingRefresh] = useState<boolean>(false);
+  const [loadingNextPage, setLoadingNextPage] = useState<boolean>(false);
+  const loadingInitialContent = loading && !loadingRefresh && !loadingNextPage;
+
+  const refresh = useCallback(() => {
+    if (loadingInitialContent || loadingRefresh) {
+      return;
+    }
+
+    setLoadingRefresh(true);
+    refetch()
+      .catch(err => Configuration.errorNotifier("Refetch Error", err))
+      .finally(() => {
+        if (isMounted.current) {
+          setLoadingRefresh(false);
+        }
+      });
+  }, [
+    loadingInitialContent,
+    loadingRefresh,
+    setLoadingRefresh,
+    refetch,
+    isMounted,
+  ]);
+
+  const nextPage = useCallback(() => {
+    if (loadingInitialContent || loadingRefresh || loadingNextPage) {
+      return;
+    }
+
+    const pageInfo = getCollectionByPath(data)?.pageInfo;
+
+    if (!pageInfo || !pageInfo.hasNextPage) {
+      return;
+    }
+
+    setLoadingNextPage(true);
+
+    fetchMore({
+      variables: {
+        cursor: pageInfo.endCursor,
+      },
+      updateQuery: (prev, { fetchMoreResult }) =>
+        fetchMoreUpdateQueryHandler(prev, fetchMoreResult, getCollectionByPath),
+    })
+      // tslint:disable-next-line:no-unsafe-any
+      .catch(err => Configuration.errorNotifier("FetchMore Error", err))
+      .finally(() => {
+        if (isMounted.current) {
+          setLoadingNextPage(false);
+        }
+      });
+  }, [
+    data,
+    loadingInitialContent,
+    loadingRefresh,
+    fetchMore,
+    loadingNextPage,
+    setLoadingNextPage,
+    getCollectionByPath,
+    isMounted,
+  ]);
+
+  useEffect(
+    () => {
+      if (subscription == undefined) return;
+
+      const subscriptionOptions = subscription.options || {};
+
+      return subscribeToMore<TSubscription>({
+        ...subscriptionOptions,
+        document: subscription.document,
+        updateQuery: (prev, { subscriptionData }) =>
+          subscribeToMoreHandler(
+            prev,
+            getCollectionByPath,
+            subscriptionData?.data,
+            subscription.getNodeByPath,
+          ),
+        onError: err =>
+          Configuration.errorNotifier("Subscribe to More Error", err),
+      });
+    },
+    // Disabling this linter so we can force this only run once. If we didn't
+    // do this we would need to ensure subscription, subscribeToMore, and getNodeByPath
+    // all use useCallback.
+    [],
+  );
+
+  return {
+    data,
+    error,
+    refresh,
+    loadingRefresh,
+    nextPage,
+    loadingNextPage,
+    loadingInitialContent,
+  };
+}
+
+/**
+ * The following method uses an `output` variable, and indirectly modifies it through the `outputCollection` variable.
```

Comment:
Does this chunk of comments matter to the end users of this hook? If so I'd add it to the mdx, as we try to surface these kinds of trade-offs and considerations where possible ([example](https://atlantis.getjobber.com/components/disclosure#notes))

Reply:
@chris-at-jobber  These comments are more about maintenance in the future and explaining decisions made. They do not matter to the end users of this hook :)